### PR TITLE
Correct tooltip when no battery

### LINKED ISSUE #172: Inaccurate tooltip on desktop
## What Happened

On desktop where there is no battery, we see the tooltip
```
Battery: 0%
Middle-click to toggle percentage
```

## Expected Behavior

Since there is no battery and no percentage, none of this information makes sense. We need a string to describe attached devices' batteries

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/96840482-inaccurate-tooltip-on-desktop?utm_campaign=plugin&utm_content=tracker%2F60466177&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60466177&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jeremypw**
Fixes #172
* Handle failure of `upower.get_display_device ()`
* Search for power supply amongst device list if necessary
* Always connect to change in display_device property
* Modify tooltip accordingly

### CODE REVIEW

**Comment by jeremypw** [comment]
This only deals with the case where there are no other battery devices available to show in the icon or tooltip - but that is also the case for master.  A subsequent PR can deal with that.

**Review by danrabbit — CHANGES_REQUESTED** [review]
With this branch, on desktop, I get the following:
```
(null)
using an external power source
```

Which I'm not sure that the secondary text here is really useful because a desktop will always be using an external power source. And for a laptop, I think this implies that the battery is missing (otherwise it would show charge status) which is an error state.

File: src/Services/Device.vala [hunk 1]
```diff
@@ -86,7 +86,7 @@ public class Power.Services.Device : Object {
                 case PEN:
                     return _("Pen");
                 case LINE_POWER:
-                case UNKNOWN:
+                    return _("Line Power");
```
[the remark this answers is not in the document]
Comment (leonardo-lemos) [reply src/Services/Device.vala:89]: "Plugged in" sounds more natural for me, I think it's easier to translate.

[the remark this answers is not in the document]
Comment (jeremypw) [reply src/Services/Device.vala:89]: OK, I'll change that.

**Comment by jeremypw** [comment]
@danrabbit OK, it looks like we need to provide a fallback name for the device when it is null. Would "Plugged in" be appropriate?  Should the secondary text be omitted in this case?

**Comment by danrabbit** [comment]
@jeremypw this is what I meant before about when we're plugged in that we're not really indicating the power status of the device, we're indicating the charge/discharge state of attached devices and/or peripherals

**Comment by jeremypw** [comment]
The tooltip now just shows "Plugged in" instead of "Line Power".  I have not been able to test PR this with devices other than the built_in battery and the AC power supply on my laptop but if the UPower interface does not return a `display_device` then it tries to choose a suitable one from the list of devices returned by UPower, prioritizing batteries over power supplies.

No second line is shown for non-battery devices.  No tooltip is shown at all if the display_device device type is null.

**Comment by danrabbit** [comment]
Hrm, I'm not really sure this is an improvement on desktop. Now, it feels like it's just picking a "random" device to represent battery. I have both a wireless mouse and keyboard plugged in, but for some reason the keyboard is the display device. That feels like a regression from the previous behavior

**Comment by jeremypw** [comment]
@danrabbit OK, so would it be better to not try to second-guess `upower.get_display_device ()` and just suppress the tooltip if that function returns `null`?

**Comment by jeremypw** [comment]
Now when the display device is `null` and a backlight is present, the tooltip shows info related to the screen brightness (which matches the icon used) otherwise there is no tooltip.

**Comment by jeremypw** [comment]
@danrabbit Is this any better?  I am not really clear on what should be displayed (if anything) when the display device  returned by upower is `null`.

**Review by danrabbit — APPROVED** [review]
This all makes sense to me :) Thanks for being patient and working this out!

**Pull request merged**